llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,386 @@
+  LLVM_DEBUG(dbgs() << "WAW Reg Renaming BasicBlock "; MBB->dump();
+             dbgs() << "\n");
+  bool Modified = false;
+  // Add the used physical registers one machine instruction at a time.
+  // This vector is used to determine, if a physical register has already been
+  // defined in the machine basic block.
+  BitVector UsedPhysRegs(TRI->getNumRegs());
+
+  // Get a list of registers, that are not allowed as a replacement register.
+  // This list gets updated with the newly replaced physical register, so that
+  // this pass does not introduce WAW dependencies.
+  BitVector BlockedPhysRegs = getDefinedPhysRegs(MBB);
+
+  // Collect all the virtual registers that have at least a copy instruction
+  // that defines them. Subregisters may contain constants that may be shared
+  // across different virtual registers. Renaming would reintroduce unnecessary
+  // copies, if physical registers are shared. Also do not rename copies, since
+  // they could be removed in a later pass.
+  SmallVector<Register, 16> VRegWithCopies = getVRegWithCopies(*MBB);
+
+  std::map<Register, const MachineInstr *> LastVRegDef = getLastVRegDef(*MBB);
+
+  for (const MachineInstr &MI : *MBB) {
+
+    // Identity copies will be removed in a later pass, therefore, these are not
+    // real defines of a physical register
+    if (isIdentityCopy(MI))
+      continue;
+
+    for (const MachineOperand &MO : MI.defs()) {
+
+      Register Reg = MO.getReg();
+      if (!Reg.isVirtual())
+        continue;
+      if (VRM->hasRequiredPhys(Reg))
+        continue;
+      if (MO.isTied())
+        continue;
+      // wait until the virtual register is fully defined to check if it is a
+      // replacement candidate
+      if (LastVRegDef[Reg] != &MI)
+        continue;
+
+      if (isWorthRenaming(Reg, UsedPhysRegs, VRegWithCopies) &&
+          replaceReg(Reg, BlockedPhysRegs)) {
+
+        LLVM_DEBUG(dbgs() << MI);
+        Modified = true;
+
+      } else {
+        // Keep track of already visited physical registers.
+        // Incrementally add the encountered physical registers, so that a
+        // second occurrence of a physical register can trigger the register
+        // rewriting
+        // Blocked registers for the replacement are recorded in
+        // BlockedPhysRegs. Initially all the used physical registers
+        // from the MBB are blocked, so that replacements do not introduce WAW
+        // dependencies. Additionally, replaced registers are already blocked in
+        // BlockedPhysRegs, so that an additional replacement will not cause a
+        // WAW, which this pass is trying to remove.
+        UsedPhysRegs[VRM->getPhys(Reg)] = true;
+      }
+    }
+  }
+
+  return Modified;
+}
+
+bool AIEWawRegRewriter::isWorthRenaming(
+    const Register &Reg, const BitVector &UsedPhysRegs,
+    const SmallVector<Register, 16> &VRegWithCopies) const {
+  assert(Reg.isVirtual() && MRI->getRegClass(Reg));
```

Comment:
By construction, virtual registers must have a register class at this stage or the Machine IR is broken.

Reply:
removed that assert

Reply:
removed that assertion